<a href="https://colab.research.google.com/github/sergioGarcia91/Introductorio-Python-3/blob/main/Python_02b_NumPy_Funciones.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Python 02a: NumPy - Funciones

> *Ser tan rápidos como el más lento,\
y ser tan lentos como el más rápido.*

**Autor:** Sergio Andrés García Arias  
**Versión 01:** Diciembre 2023

# Introducción

NumPy es una biblioteca en Python que proporciona múltiples herramientas para la manipulación de arrays y operaciones matemáticas avanzadas. A continuación, se enumeran algunas de las funciones más destacadas, junto con un enlace a su documentación respectiva, que podrían ser de tu interés:

- Creación de Arrays
  - `np.array()`: [Crear un array](https://numpy.org/doc/stable/reference/generated/numpy.array.html)
  - `np.zeros()`, `np.ones()`: [Arrays de ceros o unos](https://numpy.org/doc/stable/reference/generated/numpy.zeros.html)
  - `np.arange()`: [Crear un array con valores espaciados uniformemente en un rango](https://numpy.org/doc/stable/reference/generated/numpy.arange.html)
  - `np.linspace()`: [Crear un array con valores espaciados de manera lineal](https://numpy.org/doc/stable/reference/generated/numpy.linspace.html)
  - `np.logspace()`: [Crear un array con valores espaciados de manera logarítmica](https://numpy.org/doc/stable/reference/generated/numpy.logspace.html)

- Operaciones Matemáticas
  - `np.add()`, `np.subtract()`, `np.multiply()`, `np.divide()`, `np.sin()`, `np.cos()`, `np.exp()`: [Operaciones aritméticas entre arrays](https://numpy.org/doc/stable/reference/routines.math.html)
  - `np.sum()`, `np.mean()`, `np.max()`, `np.min()`: [Funciones de estadística](https://numpy.org/doc/stable/reference/routines.statistics.html)

- Manipulación de Arrays
  - `np.reshape()`: [Cambiar la forma de un array](https://numpy.org/doc/stable/reference/generated/numpy.reshape.html)
  - `np.concatenate()`: [Concatenar dos o más arrays](https://numpy.org/doc/stable/reference/generated/numpy.concatenate.html)
  - `np.transpose()`: [Transponer un array](https://numpy.org/doc/stable/reference/generated/numpy.transpose.html)
  - `np.append()`: [Concatenar valores al final de un array](https://numpy.org/doc/stable/reference/generated/numpy.append.html)

- Operaciones de Álgebra Lineal
  - `np.dot()`: [Producto punto entre dos arrays](https://numpy.org/doc/stable/reference/generated/numpy.dot.html)
  - `np.linalg.inv()`: [Inversa de una matriz](https://numpy.org/doc/stable/reference/generated/numpy.linalg.inv.html)
  - `np.linalg.det()`: [Determinante de una matriz](https://numpy.org/doc/stable/reference/generated/numpy.linalg.det.html)

- Generación de Números Aleatorios
  - `np.random.rand()`, `np.random.randn()`: [Generación de números aleatorios](https://numpy.org/doc/stable/reference/random/generated/numpy.random.rand.html)

- Funciones para Trabajar con Números Complejos
  - `np.real()`, `np.imag()`: [Parte real e imaginaria de un número complejo](https://numpy.org/doc/stable/reference/generated/numpy.real.html)


# Inicio

In [ ]:
import numpy as np

Siempre es bueno revisar la documentación de las librerías que se estén empleando. Para crear un array usando `numpy.arange()`, podemos recurrir a dicha [documentación](https://numpy.org/doc/stable/reference/generated/numpy.arange.html).

<center><img src='https://github.com/sergioGarcia91/Introductorio-Python-3/blob/main/imagen_NumpyArange.png?raw=true' width=1000 /></center>

Algunas librerías ofrecen recomendaciones de otras posibles funciones de interés junto con algunos ejemplos de cómo se usa la función actual de interés.

<center><img src='https://github.com/sergioGarcia91/Introductorio-Python-3/blob/main/imagen_NumpyArange_ejemplo.png?raw=true' width=1000 /></center>




Revisando la documentacion, el uso de la funcion `numpy.arange()` sería de manera general como:

```python
mi_array_a_crear = np.arange(inicio, fin, pasos)
```


In [ ]:
# Vamos a crear un array con np.arange()
array_arange = np.arange(0, 10, 2)
# Nos va generar un array del 0 al 10 (sin incluirlo) con paso 2
print("Array creado con np.arange():", array_arange)

# Y con np.array() vamos a crear otro
array_array = np.array([1, 3, 5, 7, 9])  # Lista de números convertida a array
print("Array creado con np.array():", array_array)


Array creado con np.arange(): [0 2 4 6 8]
Array creado con np.array(): [1 3 5 7 9]


In [ ]:
# Podemos conocer la forma de nuestros arrays con np.shape()
print('Forma del array_arange con np.shape(): ', np.shape(array_arange))
# O tambien con el atributo .shape
print('Forma del array_arange el atributo .shape: ', array_arange.shape)
# Ambas salidas nos da (5,) o cinco elementos

Forma del array_arange con np.shape():  (5,)
Forma del array_arange el atributo .shape:  (5,)


In [ ]:
# Podemos realizar algunas operaciones aritméticas
multiplicar = array_arange * 2
restar = array_arange - 3
suma = array_arange + 4
potencia = array_arange ** 2

print('Original: ', array_arange)
print('Multiplicar: ', multiplicar)
print('Restar: ', restar)
print('Sumar: ', suma)
print('Elevar a una potencia: ', potencia)

Original:  [0 2 4 6 8]
Multiplicar:  [ 0  4  8 12 16]
Restar:  [-3 -1  1  3  5]
Sumar:  [ 4  6  8 10 12]
Elevar a una potencia:  [ 0  4 16 36 64]


In [ ]:
# vamos a reutilizar las variables y a operar entre los dos arrays
multiplicar = array_arange * array_array
restar = array_arange - array_array
suma = array_arange + array_array
potencia = array_arange ** array_array

print('Original arange: ', array_arange)
print('Original array: ', array_array)
print('Multiplicar: ', multiplicar)
print('Restar: ', restar)
print('Sumar: ', suma)
print('Elevar a una potencia: ', potencia)

Original arange:  [0 2 4 6 8]
Original array:  [1 3 5 7 9]
Multiplicar:  [ 0  6 20 42 72]
Restar:  [-1 -1 -1 -1 -1]
Sumar:  [ 1  5  9 13 17]
Elevar a una potencia:  [        0         8      1024    279936 134217728]


- Suma de Matrices

La suma de dos matrices del mismo tamaño se realiza sumando los elementos en sus correspondientes posiciones.

$
\begin{bmatrix}
a_{11} & a_{12} \\
a_{21} & a_{22} \\
\end{bmatrix}
+
\begin{bmatrix}
b_{11} & b_{12} \\
b_{21} & b_{22} \\
\end{bmatrix}
=
\begin{bmatrix}
a_{11} + b_{11} & a_{12} + b_{12} \\
a_{21} + b_{21} & a_{22} + b_{22} \\
\end{bmatrix}
$

- Resta de Matrices

La resta de dos matrices del mismo tamaño se realiza restando los elementos en sus correspondientes posiciones.

$
\begin{bmatrix}
a_{11} & a_{12} \\
a_{21} & a_{22} \\
\end{bmatrix}
-
\begin{bmatrix}
b_{11} & b_{12} \\
b_{21} & b_{22} \\
\end{bmatrix}
=
\begin{bmatrix}
a_{11} - b_{11} & a_{12} - b_{12} \\
a_{21} - b_{21} & a_{22} - b_{22} \\
\end{bmatrix}
$

---
Pero parece que los resultados de los arrays de NumPy para la multiplicación `*` y la potencia `**` realizan la respectiva operación elemento a elemento. Esto es algo que para la multiplicación matricial no es correcto.

En NumPy, al utilizar el operador `*` o `**` en arrays, se realiza la operación elemento a elemento, y no la multiplicación matricial tradicional. Para llevar a cabo la multiplicación matricial en NumPy, es necesario utilizar `np.dot()`.



In [ ]:
# vamos a multiplicar de manera matricial array_arange con array_array
# si recordamos la multiplicación entre matrices debe
# conservar las mismas en las columnas de la primera y las filas de la segunda
# (m,n) x (n,l)
# para ello podemos usar .reshape (5, 1)x(1,5)
np.dot(array_arange.reshape(5, 1) , array_array.reshape(1,5))

array([[ 0,  0,  0,  0,  0],
       [ 2,  6, 10, 14, 18],
       [ 4, 12, 20, 28, 36],
       [ 6, 18, 30, 42, 54],
       [ 8, 24, 40, 56, 72]])

$
\begin{bmatrix}
0 \\
2 \\
4 \\
6 \\
8
\end{bmatrix}
\times
\begin{bmatrix}
1 & 3 & 5 & 7 & 9
\end{bmatrix}
=
\begin{bmatrix}
0 \times 1 & 0 \times 3 & 0 \times 5 & 0 \times 7 & 0 \times 9 \\
2 \times 1 & 2 \times 3 & 2 \times 5 & 2 \times 7 & 2 \times 9 \\
4 \times 1 & 4 \times 3 & 4 \times 5 & 4 \times 7 & 4 \times 9 \\
6 \times 1 & 6 \times 3 & 6 \times 5 & 6 \times 7 & 6 \times 9 \\
8 \times 1 & 8 \times 3 & 8 \times 5 & 8 \times 7 & 8 \times 9
\end{bmatrix}
=
\begin{bmatrix}
0 & 0 & 0 & 0 & 0 \\
2 & 6 & 10 & 14 & 18 \\
4 & 12 & 20 & 28 & 36 \\
6 & 18 & 30 & 42 & 54 \\
8 & 24 & 40 & 56 & 72
\end{bmatrix}
$


In [ ]:
# para ello podemos usar .reshape (1, 5)x(5, 1)
np.dot(array_arange.reshape(1, 5) , array_array.reshape(5, 1))

array([[140]])

$
\begin{bmatrix}
0 & 2 & 4 & 6 & 8
\end{bmatrix}
\times
\begin{bmatrix}
1 \\
3 \\
5 \\
7 \\
9
\end{bmatrix}
=
\begin{bmatrix}
(0 \times 1) + (2 \times 3) + (4 \times 5) + (6 \times 7) + (8 \times 9)
\end{bmatrix}
=
\begin{bmatrix}
140
\end{bmatrix}
$


# Broadcasting

El [Broadcasting](https://numpy.org/doc/stable/user/basics.broadcasting.html) nos permite realizar operaciones entre arrays de diferentes formas y tamaños. NumPy maneja arrays con formas diferentes durante operaciones aritméticas, como hemos podido inferir en nuestro trabajo con vectores. Ahora:

- El array más pequeño se `broadcastea` a lo largo del array más grande para que tengan formas compatibles.


In [ ]:
# Crearemos un array con np.linespace
# que inicia en 2 y termine en 3
# con un total de 6 numeros
array_linspace = np.linspace(2, 3, 6)
array_linspace # nuestro array con 6 elementos equidistantes

array([2. , 2.2, 2.4, 2.6, 2.8, 3. ])

In [ ]:
array_linspace.shape # verificamos que tenga 6 elementos

(6,)

In [ ]:
# Ahora vamos a crear una array de solo unos de con 12 elementos
# con una forma de 3 filas y 4 columnas
array_ones = np.ones( (3, 4) ) #la forma ingresa como una pareja en una tupla
array_ones # debe contener solo 1.

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

In [ ]:
array_ones.shape # verificamos la forma

(3, 4)

Para ejemplificar el broadcasting, lo haremos con el operador `*`, ya que hemos observado que opera al igual que la suma `+` y la resta `-`, elemento a elemento.

In [ ]:
# procedemos primero convirtiendo a una forma de 3,2
# nuestro
array_linspace_3x2 = np.reshape(array_linspace, (3,2))
print('Shape o forma de array_linspace_3x2: ', array_linspace_3x2.shape)
print('Shape o forma de array_ones: ', array_ones.shape)
# ahora vamos a multiplicarlo con nuestro array de unos
# no lo vamos a guardar en una variable,
# solo queremos ver el resultado
array_ones * array_linspace_3x2

Shape o forma de array_linspace_3x2:  (3, 2)
Shape o forma de array_ones:  (3, 4)


ValueError: ignored

Nos surge un error:

```python
ValueError: operands could not be broadcast together with shapes (3,4) (3,2)
```

Aunque el broadcasting es una herramienta útil para operar entre arrays, no siempre es tan sencillo. Por esta razón, se recomienda evitar el uso excesivo de esta característica de NumPy.


In [ ]:
# Vamos solo a tomar la primera colunma de array_linspace_3x2
# y a serciorarnos que tenga la forma (3,1)

columna1_array_linspace_3x2 = array_linspace_3x2[:,0].reshape(3,1)

print('Shape o forma de columna1_array_linspace_3x2: ',
      columna1_array_linspace_3x2.shape)
print(columna1_array_linspace_3x2)

print('\n')
print('Shape o forma de array_ones: ', array_ones.shape)
print(array_ones)
print('\n')

array_ones * columna1_array_linspace_3x2

Shape o forma de columna1_array_linspace_3x2:  (3, 1)
[[2. ]
 [2.4]
 [2.8]]


Shape o forma de array_ones:  (3, 4)
[[1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]]




array([[2. , 2. , 2. , 2. ],
       [2.4, 2.4, 2.4, 2.4],
       [2.8, 2.8, 2.8, 2.8]])

In [ ]:
# Ahora con una fila
# solo voy a tomar 4 valores
# y me confirmo una forma (1,4)
fila_array_linspace = array_linspace[0:4].reshape(1,4)

print('Shape o forma de fila_array_linspace: ',
      fila_array_linspace.shape)
print(fila_array_linspace)

print('\n')
print('Shape o forma de array_ones: ', array_ones.shape)
print(array_ones)
print('\n')

array_ones * fila_array_linspace

Shape o forma de fila_array_linspace:  (1, 4)
[[2.  2.2 2.4 2.6]]


Shape o forma de array_ones:  (3, 4)
[[1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]]




array([[2. , 2.2, 2.4, 2.6],
       [2. , 2.2, 2.4, 2.6],
       [2. , 2.2, 2.4, 2.6]])

In [ ]:
# Si tomo solo 2 elementos de array_linspace
array_linspace_2elementos = array_linspace[0:2].reshape(1,2)

print('Shape o forma de array_linspace_2elementos: ',
      array_linspace_2elementos.shape)
print(array_linspace_2elementos)

print('\n')
print('Shape o forma de array_ones: ', array_ones.shape)
print(array_ones)
print('\n')

array_ones * array_linspace_2elementos # Deberia salir error

Shape o forma de array_linspace_2elementos:  (1, 2)
[[2.  2.2]]


Shape o forma de array_ones:  (3, 4)
[[1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]]




ValueError: ignored

Hemos explorado varias características fundamentales de `NumPy`, desde la creación de arrays y matrices hasta operaciones básicas como la suma y resta. También hemos introducido el concepto de broadcasting, que facilita la realización de operaciones entre arrays de diferentes formas.

Aunque NumPy ofrece potentes capacidades para el manejo de datos numéricos en Python, es importante tener en cuenta las limitaciones y comprender el uso adecuado de sus funciones para evitar posibles errores. Estas habilidades son fundamentales para cualquier persona que desee trabajar en el análisis de datos, aprendizaje automático u otras disciplinas relacionadas con la ciencia de datos.

Una consideración adicional clave es que NumPy sirve como base para numerosas bibliotecas en el ecosistema de Python, lo que significa que muchas de estas bibliotecas heredan y aprovechan sus funcionalidades.

Algunas bibliotecas populares son:
- [Pandas](https://pandas.pydata.org) para manipulación de datos.
- [Matplotlib](https://matplotlib.org) para visualización de información.
- [Scikit-learn](https://scikit-learn.org/stable/index.html) para aprendizaje automático.

# Fin